File: fixtures/lua/src/operations.lua
local errors = require("src.errors")

local operations = {}

--- Return the sum of a and b.
--- @param a number
--- @param b number
--- @return number
function operations.add(a, b)
    return a + b
end

--- Return the difference of a and b.
--- @param a number
--- @param b number
--- @return number
function operations.subtract(a, b)
    return a - b
end

--- Return the product of a and b.
--- @param a number
--- @param b number
--- @return number
function operations.multiply(a, b)
    return a * b
end

--- Return the quotient of a and b.
--- Raises an error if b is zero.
--- @param a number
--- @param b number
--- @return number
function operations.divide(a, b)
    if b == 0 then
        error(errors.division_by_zero())
    end
    return a / b
end

return operations

File: fixtures/haskell/src/Operations.hs
module Operations
  ( add,
    subtract,
    multiply,
    divide,
  )
where

import Errors (CalculatorError (..))

-- | Returns the sum of two numbers.
add :: Double -> Double -> Double
add a b = a + b

-- | Returns the difference of two numbers.
subtract :: Double -> Double -> Double
subtract a b = a - b

-- | Returns the product of two numbers.
multiply :: Double -> Double -> Double
multiply a b = a * b

-- | Returns the quotient of two numbers.
-- Returns an error if the divisor is zero.
divide :: Double -> Double -> Either CalculatorError Double
divide _ 0.0 = Left DivisionByZero
divide a b = Right (a / b)
